File: Medium/767. Reorganize String/767. Reorganize String.js
/**
 * @param {string} s
 * @return {string}
 */
var reorganizeString = function (s) {
    const charCount = new Array(26).fill(0);

    for (const char of s) {
        const index = char.charCodeAt(0) - 'a'.charCodeAt(0);
        charCount[index]++;
    }

    const maxCount = Math.max(...charCount);
    if (maxCount > Math.floor((s.length + 1) / 2)) {
        return '';
    }

    const result = new Array(s.length);
    let evenIndex = 0,
        oddIndex = 1;
    const halfLength = Math.floor(s.length / 2);

    for (let i = 0; i < 26; i++) {
        const char = String.fromCharCode('a'.charCodeAt(0) + i);

        while (
            charCount[i] > 0 &&
            charCount[i] <= halfLength &&
            oddIndex < s.length
        ) {
            result[oddIndex] = char;
            charCount[i]--;
            oddIndex += 2;
        }

        while (charCount[i] > 0) {
            result[evenIndex] = char;
            charCount[i]--;
            evenIndex += 2;
        }
    }

    return result.join('');
};

File: Medium/767. Reorganize String/767. Reorganize String.java
class Solution {
    public String reorganizeString(String s) {
        int[] charCount = new int[26];
        
        for (char c : s.toCharArray()) {
            charCount[c - 'a']++;
        }
        
        int maxCount = Arrays.stream(charCount).max().getAsInt();
        if (maxCount > (s.length() + 1) / 2) {
            return "";
        }
        
        char[] result = new char[s.length()];
        int evenIndex = 0, oddIndex = 1;
        int halfLength = s.length() / 2;
        
        for (int i = 0; i < 26; i++) {
            char c = (char) ('a' + i);
            
            while (charCount[i] > 0 && charCount[i] <= halfLength && oddIndex < s.length()) {
                result[oddIndex] = c;
                charCount[i]--;
                oddIndex += 2;
            }
            
            while (charCount[i] > 0) {
                result[evenIndex] = c;
                charCount[i]--;
                evenIndex += 2;
            }
        }
        
        return new String(result);
    }
}